### PR TITLE
 MIPS: Fix asm constraints "f" and "r" for softfloat

### LINKED ISSUE #64241: Clang MIPS: float register constraints for inline asm deviate from gcc
clang and gcc seem to have different behavior of floating point constraints for inline asm targetting MIPS. Consider this simple example

```c
// gcc compiles this but clang does not
void read_float(float* p) {
    float result = *p;
    __asm__("" ::"r"(result));
}
```
The following code compiles fine on gcc when targeting MIPS, but clang gives a compile error `couldn't allocate input reg for constraint 'r'`

Clang requires the register constraint here to be `"f"` not `"r"`. However applying an `"f"` constraint, causes gcc to fail compilation with the error `error: impossible constraint in 'asm'` while clang compiles the code fine. 

```c
// clang compiles this but gcc does not
void read_float(float* p) {
    float result = *p;
    __asm__("" ::"f"(result));
}
```

Here is a simple demo of the issue: https://godbolt.org/z/WGWxd3ejo
Test with "mips gcc 13.1.0" and "mips clang 16.0.0"

(This bug was discovered as part of the wasm2c project https://github.com/WebAssembly/wabt/issues/2266)

### PR DESCRIPTION
**Pull request by wzssyqa**
This include 2 fixes:
        1. Disallow 'f' for softfloat.
        2. Allow 'r' for softfloat.

Currently, 'f' is accpeted by clang, then LLVM meets an internal error.

'r' is rejected by LLVM by:  couldn't allocate input reg for constraint 'r'.

Fixes: #64241, #63632

### CODE REVIEW

File: clang/test/Driver/mips-float.c [hunk 1]
```diff
@@ -102,3 +102,50 @@
...
 // CHECK-ABI-SOFT-MIPS16: "-msoft-float"
 // CHECK-ABI-SOFT-MIPS16: "-mfloat-abi" "soft"
+
+/// On MIPS, don't accept constraint "f" for soft-float.
+// RUN: not %clang -S %s -o %t.s 2>&1 \
+// RUN:     -target mips-linux-gnu -msoft-float \
+// RUN:     -DSOFT_FLOAT_NO_CONSTRAINT_F \
+// RUN:   | FileCheck --check-prefix=CHECK-SOFTFLOAT-ASM-NO-F %s
+// CHECK-SOFTFLOAT-ASM-NO-F: error: invalid input constraint 'f' in asm
+
```
Comment (MaskRay) [inline clang/test/Driver/mips-float.c:112]: test/Driver should only test how the driver passes options to cc1.

For inline assembly, you usually test that clangCodeGen correctly generates LLVM IR and the backend can lower asm. See `#77886` (x86 "Ws") for a good example (I carefully add tests. Many early-day asm patches might not add appropriate tests.)

File: clang/test/Sema/inline-asm-validate-mips.c [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// RUN: %clang_cc1 -triple mips -target-feature +soft-float -DSOFT_FLOAT_NO_CONSTRAINT_F -fsyntax-only -verify %s
+
+#ifdef SOFT_FLOAT_NO_CONSTRAINT_F
+void read_float(float* p) {
```
Comment (MaskRay) [inline clang/test/Sema/inline-asm-validate-mips.c:4]: Can just use argument `float p`?

File: llvm/test/CodeGen/Mips/inlineasm-constraints-softfloat.ll [hunk 3]
```diff
@@ -0,0 +1,25 @@
+; RUN: llc  -march=mips < %s | FileCheck %s --check-prefix=MIPS32
+; RUN: llc  -march=mips64 < %s | FileCheck %s --check-prefix=MIPS64
+
+define dso_local void @read_double(double* nocapture noundef readonly %0) local_unnamed_addr #0 {
```
Comment (MaskRay) [inline llvm/test/CodeGen/Mips/inlineasm-constraints-softfloat.ll:4]: Newer tests should use opaque pointers (`ptr`)

**Comment by brad0** [comment]
@wzssyqa Ping.

File: clang/test/Sema/inline-asm-validate-mips.c (hunk 2 above)
Comment (wzssyqa) [reply clang/test/Sema/inline-asm-validate-mips.c]: Oh, yes. we can use `float p` here.

File: llvm/test/CodeGen/Mips/inlineasm-constraints-softfloat.ll (hunk 3 above)
Comment (wzssyqa) [reply llvm/test/CodeGen/Mips/inlineasm-constraints-softfloat.ll]: Thanks.

**Comment by brad0** [comment]
@MaskRay

File: llvm/test/CodeGen/Mips/inlineasm-constraints-softfloat.ll [hunk 4]
```diff
@@ -0,0 +1,25 @@
+; RUN: llc  -march=mips < %s | FileCheck %s --check-prefix=MIPS32
+; RUN: llc  -march=mips64 < %s | FileCheck %s --check-prefix=MIPS64
+
+define dso_local void @read_double(ptr nocapture noundef readonly %0) local_unnamed_addr #0 {
+  %2 = load double, ptr %0, align 8
+; MIPS32-LABEL: read_double:
+; MIPS32: lw      $2, 4($4)
```
Comment (MaskRay) [inline llvm/test/CodeGen/Mips/inlineasm-constraints-softfloat.ll:7]: Ensure that `lw` is indented by 2 compared with `read_double:`

File: clang/lib/Basic/Targets/Mips.h [hunk 5]
```diff
@@ -238,6 +238,9 @@ class LLVM_LIBRARY_VISIBILITY MipsTargetInfo : public TargetInfo {
     case 'd': // Equivalent to "r" unless generating MIPS16 code.
     case 'y': // Equivalent to "r", backward compatibility only.
     case 'f': // floating-point registers.
+      if (*Name == 'f' && FloatABI == SoftFloat)
```
Comment (MaskRay) [inline clang/lib/Basic/Targets/Mips.h:241]: It's better to move 'd' and 'y' below. I'll fix it.

**Review by MaskRay — APPROVED** [review]

**Pull request merged**